proto/api/v1/auth_service.proto

```diff
@@ -20,21 +21,146 @@ message UserAttribute {
+}
+
+message AddAuthMachineToMachineConfigRequest {
+    AuthMachineToMachineConfig config = 1;
+}
+
+message AddAuthMachineToMachineConfigResponse {
+    AuthMachineToMachineConfig config = 1;
+}
+
+message UpdateAuthMachineToMachineConfigRequest {
+    AuthMachineToMachineConfig config = 1;
+}
+
+message UpdateAuthMachineToMachineConfigResponse {
+    AuthMachineToMachineConfig config = 1;
+}
+
+message ExchangeAuthMachineToMachineTokenRequest {
+    // Identity token that is supposed to be exchanged.
+    string id_token = 1;
+}
 
-    storage.AuthProvider   auth_provider   = 5;
-    storage.UserInfo       user_info       = 6;
-    repeated UserAttribute user_attributes = 7;
+message ExchangeAuthMachineToMachineTokenResponse {
+    // The exchanged access token.
+    string access_token = 1;
 }
 
 service AuthService {
+    // GetAuthStatus returns the status for the current client.
     rpc GetAuthStatus(Empty) returns (AuthStatus) {
         option (google.api.http) = {
             get: "/v1/auth/status"
         };
     }
+
+    // ListAuthMachineToMachineConfigs lists the available auth machine to machine configs.
+    rpc ListAuthMachineToMachineConfigs(Empty) returns (ListAuthMachineToMachineConfigResponse) {
+        option (google.api.http) = {
+            get: "/v1/auth/m2m"
+        };
+    }
+
+    // GetAuthMachineToMachineConfig retrieves the specific auth machine to machine config.
+    rpc GetAuthMachineToMachineConfig(ResourceByID) returns (GetAuthMachineToMachineConfigResponse) {
+        option (google.api.http) = {
+            get: "/v1/auth/m2m/{id}"
+        };
+    }
+
+    // AddAuthMachineToMachineConfig creates a new auth machine to machine config.
+    rpc AddAuthMachineToMachineConfig(AddAuthMachineToMachineConfigRequest) returns (AddAuthMachineToMachineConfigResponse) {
+        option (google.api.http) = {
+            post: "/v1/auth/m2m"
+            body: "*"
+        };
+    }
+
+    // UpdateAuthMachineToMachineConfig updates an existing auth machine to machine config.
+    rpc UpdateAuthMachineToMachineConfig(UpdateAuthMachineToMachineConfigRequest) returns (UpdateAuthMachineToMachineConfigResponse) {
+        option (google.api.http) = {
+            put: "/v1/auth/m2m"
+            body: "*"
+        };
+    }
+
+    // DeleteAuthMachineToMachineConfig deletes the specific auth machine to machine config.
+    rpc DeleteAuthMachineToMachineConfig(ResourceByID) returns (Empty) {
+        option (google.api.http) = {
+            delete: "/v1/auth/m2m/{id}"
+        };
+    }
+
+    // ExchangeAuthMachineToMachineToken exchanges a given identity token for a Central access token based on
+    // configured auth machine to machine configs.
+    rpc ExchangeAuthMachineToMachineToken(ExchangeAuthMachineToMachineTokenRequest) returns (ExchangeAuthMachineToMachineTokenResponse) {
+        option (google.api.http) = {
+            post: "/v1/auth/m2m/exchange"
+            body: "*"
+        };
+    }
```

Comment:
No not at all, this would be a first.

I'm also not sure about _which_ types of errors would make sense here: the error will be `errox.NoCredentials` (i.e. 401). So, we could maybe document it, but I'm not entirely sure of its value being in this written form if it's not encompassed in the generated swagger JSON as this is probably the way most people consume our APIs (I haven't found _quickly_ how such documentation would be done for gRPC specific stuff yet).

Reply:
While I agree with Misha's point, making it happen will require not just work but also a mental shift. Currently, we don't have a clear separation of responsibilities between store/datastore/service hence error _types_ (or _classes_) which determine gRPC and HTTP status codes are generated on all layers. To spice this up, interceptors in the request processing pipeline can short-circuit and emit their own error classes as well.

Given the above I'm afraid documenting possible status codes will do more harm than good: saying nothing is better than lying.

Reply:
I wasn't clear enough. My ask here was broader/different than HTTP status codes.

When I as a developer call `ExchangeAuthMachineToMachineToken` and it does not succeed, I want to know **why**. Look: I **really** do.
Many things can go wrong when I integrate with `ExchangeAuthMachineToMachineToken` for the first time.
- My `id_token` may be expired.
- My `id_token` may not resolve to a valid M2M config.
- The issuer of my `id_token` may fail validation according to what's in M2M config.
- ... You definitely know more common cases that can go wrong.

As a developer I would be very interested in having a good textual description returned so that I have a direction how to fix my code. Accurate HTTP status code is less important, it's nice-to-have.

Perhaps this PR is a bit too early to dive into that, because the implementation is added in <https://github.com/stackrox/stackrox/pull/8176>.

Reply:
Let's move this one to the follow-up PR, my initial thoughts on this:

Given that this is an `auth` API, and is unexposed, I'm hesitant to return _too_ much information about failures (e.g. specifics about m2m configs).

The result of this is relatively clear to me: either way, you are not going to be seen as authorized, so the HTTP status code should be 401 for _all_ failure cases, but the message may be different.

But, let's move this discussion to the follow-up PR.